ttnn/cpp/ttnn/operations/normalization/layernorm/device/kernels/compute/layernorm_sharded_post_allgather.cpp

```diff
@@ -0,0 +1,333 @@
+    const bool is_second_stage_reader                       = is_allgather_worker ? get_arg_val<uint32_t>(3) == 1 : false;
+    const uint32_t num_distributed_blocks                   = is_allgather_worker ? get_arg_val<uint32_t>(4) : 0;
+
+    uint32_t num_blocks_reduce;
+    if (is_second_stage_reader) {
+        num_blocks_reduce = num_blocks_first_stage + num_blocks_second_stage - 1;
+    } else {
+        num_blocks_reduce = num_blocks_first_stage;
+    }
+
+    bool enable_sqrt;
+    if (use_two_stage_reduce and not is_second_stage_reader) {
+        enable_sqrt = false;
+    } else {
+        enable_sqrt = true;
+    }
+
+    constexpr uint32_t dst0 = 0;
+    constexpr uint32_t dst1 = 1;
+    constexpr uint32_t scaler0 = 0;
+
+    constexpr uint32_t cb_in0 = tt::CB::c_in0;
+    constexpr uint32_t cb_in1 = tt::CB::c_in1;
+    constexpr uint32_t cb_eps = tt::CB::c_in3;
+    constexpr uint32_t cb_scaler_global = tt::CB::c_in4;
+    constexpr uint32_t cb_gamma = tt::CB::c_in5;
+    constexpr uint32_t cb_beta = tt::CB::c_in6;
+
+    constexpr uint32_t cb_ex = tt::CB::dataflow1; // E[x] global reduce
+    constexpr uint32_t cb_ex2 = tt::CB::dataflow4; // E[x^2]
+    constexpr uint32_t cb_stats = tt::CB::c_in7; // E[(x-E[x])^2] global reduce
+    constexpr uint32_t cb_stats_reduced = tt::CB::c_intermed4; // E[(x-E[x])^2] global reduce
+    constexpr uint32_t cb_ex_global = tt::CB::dataflow7; // E[x] global reduce
+    constexpr uint32_t cb_reciprocal = tt::CB::c_intermed3; // [E[x^2]-E[x]^2]+eps
+    constexpr uint32_t cb_fusion = tt::CB::c_intermed1; // stream gamma/beta
+    constexpr uint32_t cb_out = tt::CB::c_out0;
+    constexpr uint32_t cb_var = tt::CB::c_intermed2;
+    constexpr uint32_t cb_ex_sqr = tt::CB::c_intermed0; // E[x]^2
+
+
+    #ifdef RMSNORM
+    binary_op_init_common(cb_stats, cb_scaler_global, cb_var);
+    constexpr uint32_t stats_tiles = 1;
+    constexpr uint32_t cb_xmm = cb_in0; // x
+    #else
+    binary_op_init_common(cb_stats, cb_scaler_global, cb_stats_reduced);
+    constexpr uint32_t stats_tiles = 2;
+    constexpr uint32_t cb_xmm = tt::CB::c_intermed1; // x minus mean
+    #endif
+
+
+    // set block_h to volatile to disable automatically unroll of the loops, avoid code overflow
+    const uint32_t block_h = (block_w == 1) ? block_h_volatile : block_h_const;
+    const uint32_t subblock_w = (block_w <= 2) ? subblock_w_volatile : subblock_w_const;
+
+    int index_subblock_w_offset = 0;
+    int index_h_offset = 0;
+    int index = 0;
+
+    constexpr uint32_t cb_im = (do_gamma | do_beta) ? cb_ex_sqr : cb_out;
+    constexpr uint32_t cb_outgamma = do_beta ? cb_fusion : cb_out;
+
+
+    // global reduce, cb_ex <-- cb_ex_external, cb_ex_partial
+    if constexpr(is_allgather_worker) {
+        if (enable_sqrt) {
+            #ifdef RMSNORM
+            cb_reserve_back(cb_var, 1);
+            #else
+            cb_reserve_back(cb_stats_reduced, 1);
+            cb_reserve_back(cb_ex2, 1);
+            #endif
```

Comment:
is the stats are sharded on one core only, then is_allgather_worker here is not correct, there will be multiple all gather workers when tensor height > 1

Reply:
Yes, Stats are sharded on first core only which is the second stage all_gather_worker. Currently this kernels only support tensors height = 1 tile.